### PR TITLE
Cost-tracking: #2042, intercepting contract-call costs

### PR DESCRIPTION
**Pull request by kantai**
## Description

This PR implements #2042 -- it allows a CostTracker to intercept the cost calculation of contract-call invocations when specific targets are used.

This will be used in the cost voting system to allow STX holders and miners to introduce new "native" Clarity functions by specifying a new cost for those functions.

## Type of Change
- New feature

## Does this introduce a breaking change?

No, this is not a breaking change.

## Are documentation updates required?

No.

## Testing information

This PR includes unit tests covering the behavior of the LimitedCostTracker with contract-call interceptions.

### CODE REVIEW

File: src/vm/contexts.rs [hunk 1]
```diff
@@ -779,6 +799,23 @@ impl<'a, 'b> Environment<'a, 'b> {
...
+    pub fn run_free<F, A>(&mut self, to_run: F) -> A
+    where
+        F: FnOnce(&mut Environment) -> A,
+    {
+        let original_tracker = replace(
+            &mut self.global_context.cost_track,
+            LimitedCostTracker::new_free(),
+        );
+        // note: it is important that this method not return until original_tracker has been
+        //  restored. DO NOT use the try syntax (?).
```
Comment (jcnelson) [inline src/vm/contexts.rs:813]: Not sure what this comment is saying -- isn't that already guaranteed here by the way the code is written?  Or, is this a comment on the implementation of `to_run`?

**Comment by jcnelson** [comment]
So, issue #2042 said that this should be limited to read-only contract-calls?  I don't see any read-only checks.  Is this okay?

File: src/vm/costs/mod.rs [hunk 2]
```diff
@@ -409,6 +457,26 @@ impl CostTracker for LimitedCostTracker {
...
+        input: &[u64],
+    ) -> Result<bool> {
+        if self.free {
+            // if we're already free, no need to worry about short circuiting contract-calls
+            return Ok(false);
+        }
+        // grr, if HashMap::get didn't require Borrow, we wouldn't need this cloning.
+        let lookup_key = (contract.clone(), function.clone());
+        if let Some(cost_function) = self.contract_call_circuits.get(&lookup_key).cloned() {
+            let input_size = input.iter().fold(0, |agg, cur| agg + cur);
```
Comment (jcnelson) [inline src/vm/costs/mod.rs:473]: I guess I don't understand what `input` is here.  Is input just a list of instance characteristics?  If so, then why is _summing_ them the right thing to do?  Consider matrix multiplication as a counterexample -- the runtime would be the _product_ of the instance characteristics.  Or, am I misunderstanding what's going on here?

File: src/vm/contexts.rs (hunk 1 above)
Comment (kantai) [reply src/vm/contexts.rs:813]: Yeah, it's a note to future modifiers of that code.

File: src/vm/costs/mod.rs (hunk 2 above)
Comment (kantai) [reply src/vm/costs/mod.rs:473]: That's the size of each input argument. Ultimately, this is going to change to just be multiple arguments, and then the defined cost function can choose how to combine them.

Comment (jcnelson) [reply src/vm/costs/mod.rs:473]: Gotcha.  Is there an issue open for this?  If not, can you open one?

**Comment by kantai** [comment]
> So, issue #2042 said that this should be limited to read-only contract-calls? I don't see any read-only checks. Is this okay?

Ah, yes, this should check that the target is read-only.

File: src/vm/costs/mod.rs (hunk 2 above)
Comment (kantai) [reply src/vm/costs/mod.rs:473]: Opened: https://github.com/blockstack/stacks-blockchain/issues/2075

**Comment by kantai** [comment]
>> So, issue #2042 said that this should be limited to read-only contract-calls? I don't see any read-only checks. Is this okay?
>
> Ah, yes, this should check that the target is read-only.

Actually, looking more into this, I think we'll want to enforce this at the same time we enforce #2041. To enforce it at time-of-invocation, we'd need to look up the target function in the MARF to check whether or not it is read-only, which is already much of the cost of invocation.

File: src/vm/costs/mod.rs (hunk 2 above)
Comment (reedrosenbluth) [reply src/vm/costs/mod.rs:473]: We've actually already got this issue: https://github.com/blockstack/stacks-blockchain/issues/2047

**Review by lgalabru — APPROVED** [review]
Looks good to me, thank you @kantai!

**Comment by lgalabru** [comment]
Just asked for a round of review - merging this branch would unblock progress on https://github.com/blockstack/stacks-blockchain/pull/2109

**Review by reedrosenbluth — APPROVED** [review]

**Review by jcnelson — APPROVED** [review]
LGTM!

**Pull request merged**